### DISCUSSION TITLE
Porting a Django App to Toga

### DISCUSSION THREAD

**Discussion by gregcowell** [beeware/toga#1520] [Ideas]
I have written a Django based app for automatic generation of shift rosters (see link to video and Github repo). Many of the people who get in touch with me about the app do not work in IT and just want a version they can install on their desktop. The app also uses the Google OR-Tools Python library. I am thinking that the user interface is fairly simple so Toga may be mature enough for me to create a Windows version. Any thoughts or suggestions ?

[Video Demo](https://youtu.be/ljNMU56XjxE)
[GitHub repo](https://github.com/galojix/roster-wizard)

**Comment by freakboy3742** [comment]
I've only taken a quick look at the demo video, so it's possible I might be missing some of the finer details, but building that app in Toga certainly seems plausible. I'm guessing you'll end up with a bunch of content panels in an OptionContainer, with each of the panels managing a Table of some kind. 

From what I can see, I'm guessing you'll end up with a Table Source to manage the underlying roster data; with a single TableSource being visualised by multiple different Tables; so it would be worth getting familiar with the TableSource demo and understanding what it's doing and why. 

Other than that, the only other guidance/warning I'd give is that Toga is the general warning that Toga is still in rapid development, so it's possible you'll hit against some rough edges or missing features. However, having real world examples of usage helps to guide new feature development, so any feedback on your development experience would be welcome.

**Reply by gregcowell** [reply]
Thanks! I will give it a try and provide feedback on how it goes.

**Comment by gregcowell** [comment]
Here is some initial feedback. I have my app partially working using Toga. I went with OptionContainers as suggested. Each OptionContainer (tab) has a Box to which I add forms (lists of widgets) as required. Most of the work has been working out how to use Table Sources and integrating them with a database (via Django ORM) and working out how to structure the project. The new async functionality in Django has come in handy. I ended up using a separate Source for each of my tables as this seemed to be the most logical way to organise it. The tables in this scenario are the ones that the user sees on screen. They do not directly correspond to database tables. Overall Toga seems excellent and I plan to finish porting my Django app as soon as possible. If you interested in seeing how I have structured the code it is here: [Toga app](https://github.com/galojix/roster-wizard-desktop). I will have a go at a PR for a Multiple Selection Widget as discussed in #1556.

**Reply by freakboy3742** [reply]
Thanks for that update - looks like you've made good progress!

From what I can see, your decision to use a separate source was a good call - there might be an opportunity for an abstraction to factor out some common behavior for a "database backed source", but on the whole, it looks like you've got the idea.

If you've got any feedback on any friction you encountered - especially with the Source API - we'd love to hear it; and, of course, PRs are always welcome.

**Comment by gregcowell** [comment]
Thanks! Yes, I would definitely like to factor out the common behaviour for the database backed sources. I think I'll leave that until I understand them a lot better.

The examples have been the best source of information for the Source API so far. However, I have been struggling with working out how to use _notify() in an optimal way. For example, I have a table source that needs to be refreshed when another table source has been updated. The method I have written just clears the data source, reads the data from the database one item at a time and calls _notify("insert", ...) for each item. I suspect there is a more efficient way to do this. Perhaps I just need to dig into the source code more to understand how _notify() works.

**Reply by freakboy3742** [reply]
There's definitely room for improvement on our documentation - "documentation by example" isn't a great way of communicating. If you felt like exercising your writing muscles and contributing a topic guide for sources, that would definitely be an appreciated (and much needed) contribution.

There's almost certainly room for improvement on the API for `_notify()`; it's currently very heavily focussed on the "update a single item" use case, which is used to trigger updates in the display of a single item in the GUI. However, in the "bulk update" case, having a flood of individual notification events will be problematic if you have a lot of data. So - there's almost certainly room for improvement here.

There are two approaches I can see working. One is a separate "Everything has changed" signal; the second is to take a page from the way GUIs handle screen updates, and have a way to "suspend" updates for a moment, because you know that there's going to be a lot of changes. The approach here is to suspend updates, make the changes; and when the suspension is lifted, it's a trigger to redraw *everything* that has been marked as dirty (as opposed to N individual drawing events, which can be slow). The former will definitely be easier to implement; but the latter doesn't require making a special case of the "update all" case.

**Reply by gregcowell** [reply]
Thanks, I understand _notify() better now. 

On the subject of documentation, I think I would have found it hard to get started if you had not pointed me in the direction of OptionContainers at the start of this discussion. I am wondering why OptionContainer is not called TabContainer and OptionItem is not called TabItem but I am guessing it is because they are used for things other than creating tabs.

I'll start taking notes as I learn more about sources with the end goal of contributing a topic guide for sources. Would that  involve submitting a pull request to update data-sources.rst in the Toga repo ?

**Reply by freakboy3742** [reply]
The best explanation of the naming that I can give is that a "tab" implies a specific rendering which doesn't *necessarily* have to map to how the widget is rendered. For example, on macOS, it's not really a "Tab" in the traditional sense - it's a set of buttons that are visually distinct from the pages of content. Toga's API tries (as much as possible) to separate the conceptual idea that a GUI is trying to express (in this case, a collection of content where the currently selected content is an option for the user), from the specific rendering of that idea (using tabs).

As for a contribution - yes; that's the file you'd be updating.

**Reply by gregcowell** [reply]
I have progressed a lot further with my app and have done a lot of refactoring. I am getting closer to being able to contribute a topic guide for sources. There is one thing I am still quite puzzled about. I understand the concept of a widget adding itself to a data sources's listener list and a data source running _notify() to invoke a method on each of the widgets in the listener list. However when using a table data source I have noticed that a _notify() is not enough to get the contents of a table widget to update. I have had to remove the table widget from the box it resides in and then re-add it to the box in order for the updated table contents to show. It could be that I am just using the source incorrectly.

**Reply by freakboy3742** [reply]
It's also entirely possible you've found a bug :-) If you can narrow your case down to a simple reproduction case, open a ticket and we can investigate further.

**Reply by gregcowell** [reply]
Will do. I noticed that "Clear Table" in the table source example does not update the onscreen table contents so perhaps this is related.

**Reply by gregcowell** [reply]
I updated Toga to the latest development release and my table widgets are now  updating in response to _notify(). However I suspect the problem may have been related to an earlier revision of my code.
